### PR TITLE
add GenericRankNTensor

### LINKED ISSUE #15276: GenericRankNTensor
## Reason

When I am converting some of my AD materials, I need GenericRankTwoTensor<is_ad>

## Design

Add typedefs to MooseTypes.h

### PR DESCRIPTION
**Pull request by hugary1995**
closes #15276

### CODE REVIEW

File: modules/tensor_mechanics/src/materials/ComputeEigenstrain.C [hunk 1]
```diff
@@ -23,15 +25,21 @@ ComputeEigenstrain::validParams()
...
 }
 
+template <bool is_ad>
 void
-ComputeEigenstrain::computeQpEigenstrain()
+ComputeEigenstrainTempl<is_ad>::computeQpEigenstrain()
 {
   // Define Eigenstrain
-  _eigenstrain[_qp] = _eigen_base_tensor * _prefactor[_qp];
+  this->_eigenstrain[this->_qp] = _eigen_base_tensor * _prefactor[this->_qp];
```
Comment (hugary1995) [inline modules/tensor_mechanics/src/materials/ComputeEigenstrain.C:41]: @dschwen Do you know how to get rid of these annoying "this->" and the "this->template" above?

Comment (dschwen) [reply modules/tensor_mechanics/src/materials/ComputeEigenstrain.C:41]: With a `using` declaration. We'll basically have to bring back the `usingXXXMembers` macros that we had for the old AD classes that were templated on ComputeStage. It's not pretty, but it works.

Comment (hugary1995) [reply modules/tensor_mechanics/src/materials/ComputeEigenstrain.C:41]: Okay I see. I guess I'll just forget about merging those AD materials in this PR and only put in the generic template (without testing?)

**Comment by dschwen** [comment]
You can always put in a unit test.

**Comment by hugary1995** [comment]
@dschwen Could you please take a quick look at this? I didn't convert any of the existing materials, but this should enable users to write generic materials with RankNTensors.

**Review by dschwen — APPROVED** [review]
Yep, looks good to me.

**Pull request merged**